### PR TITLE
Greg/typescript conversion

### PR DESCRIPTION
**Pull request by GregTheGreek**
This started the conversion to typescript.

All the transition functions and helper functions have been implemented.
Tests have been written for `clamp()` and `split()`

This branch should be merged to master simply so people can contribute to the typescript conversion and implementation as a whole.

Caveat:
- Certain numbers (`uint<number>`, `bytes`...etc) have been alias to `number` for the time being until we have a proper implementation for them.

EDIT:
I added travis-ci and deleted the old non-typescript code as it was out-of-date and incomplete.

### CODE REVIEW

File: beaconChain/typescript/helpers/stateTransition.ts [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+  return []
+}
+
+/**
+ * Splits a list into split_count pieces.
+ * @param {any[]} values
+ * @param {int} splitCount
+ * @returns {any[]}
+ */
+export function split(values: any[], splitCount: int): any[] {
```
Comment (ansermino) [inline beaconChain/typescript/helpers/stateTransition.ts]: I wonder if generics might worth while here, perhaps something like this:
```
export function split<T>(values: T[], splitCount: int): T[] {
  // Returns the split ``seq`` in ``split_count`` pieces in protocol.
  const listLength = values.length;
  let array: T[] = [];
  for (let i = 0; i < splitCount; i++) {
    array.push(values.slice(
      Math.floor((listLength * i) / splitCount), Math.floor((listLength * (i + 1)) / splitCount)
    ));
  }
  return array;
}
```

I do get an error on the above, but perhaps some food for thought

File: beaconChain/typescript/interfaces/state.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  // Last cycle-boundary state recalculation
+  lastStateRecalculationSlot: uint64,
+  // Last finalized slot
+  lastFinalizedSlot: uint64,
+  // Last justified slot
+  lastJustifiedSlot: uint64,
+  // Number of consecutive justified slots
+  justifiedStreak: uint64,
+  // Committee members and their assigned shard, per slot
+  shardAndCommitteeForSlots: [[ShardAndCommittee]],
```
Comment (ansermino) [inline beaconChain/typescript/interfaces/state.ts]: Please update all the arrays to use consistent notations (ie. type[] or type[][])

**Review by ansermino — CHANGES_REQUESTED** [review]
Great work Gurg! Can't say I've given this a thorough review, feel free to merge but I will come back later tonight and take a second glance. Also -- some updates to the README would be great.

File: beaconChain/typescript/helpers/stateTransition.ts (hunk 1 above)
Comment (Mikerah) [reply beaconChain/typescript/helpers/stateTransition.ts]: +1 for generics. The use of `any` makes me squirm.

**Comment by Mikerah** [comment]
I object to the use of aliases for readability reasons. You can just have placeholders/stubs for such a use case until we get the types sorted out.

**Review by Mikerah — DISMISSED** [review]
Remove aliases. This will cause confusion in the future and will inhibit readability of our code. I will be creating an issue regarding our style guide.

**Comment by GregTheGreek** [comment]
@mikerah I have to disagree, aliases Are essentially stubs and they maintain the correct type for when we swap them out. Adding stubs ( i.e using number out right ) would mean that we are not following the spec.

**Comment by Mikerah** [comment]
I think it's misleading to have `uint<number>` and `bytes` aliased to `number`, for example. These are not the same. I think it's best to leave it empty until we (or an outside contributor) comes up with a solution.

**Comment by GregTheGreek** [comment]
Again I'll have to disagree. We can't compile the code, nor can we have complete functions without aliases. Which means we can't run tests.

**Comment by ansermino** [comment]
@Mikerah Once we finish building the uint and int types we just swap the aliases with an import. Seemed more important to move ahead with the conversion then to wait for the type implementations.

**Comment by GregTheGreek** [comment]
@Mikerah @ansermino updated.

File: beaconChain/helpers/stateTransitionHelpers.ts [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    if (x.shard === attestationData.shard) return x;
+  })[0];
+
+  // TODO Figure out why this is an error
+  // TODO implement error based on python pseudo code
+  // TODO what is ceil_div8()
+  // assert len(participation_bitfield) == ceil_div8(len(snc.committee))
+
+  const participants: int[] = snc.committee.filter((validator: uint24, index: int) => {
+    const bit = (participationBitfield[Math.floor(index / 8)] >> (7 - (index % 8))) % 2;
```
Comment (ansermino) [inline beaconChain/helpers/stateTransitionHelpers.ts]: Missing type

File: beaconChain/helpers/stateTransitionHelpers.ts [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+/**
+ * Calculate the largest integer k such that k**2 <= n.
+ * Used in reward/penalty calculations
+ * @param {int} n
+ * @returns {int}
+ */
+// TODO Can use built in JS function if available
+export function intSqrt(n: int): int {
+  let x = n;
```
Comment (ansermino) [inline beaconChain/helpers/stateTransitionHelpers.ts]: Missing types

File: beaconChain/interfaces/blocks.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+export interface BeaconBlock {
+  // Slot number
+  slot: uint64,
+  // Proposer RANDAO reveal
+  randaoReveal: hash32,
+  // Recent PoW receipt root
+  candidatePowReceiptRoot: hash32,
+  // Skip list of previous beacon block hashes
+  // i'th item is the most recent ancestor whose slot is a multiple of 2**i for i = 0, ..., 31
+  ancestorHashes: [hash32],
```
Comment (ansermino) [inline beaconChain/interfaces/blocks.ts]: Not a valid array

**Review by ansermino — COMMENTED** [review]
Please lint, and add linting to to Travis. Feel free to merge after addressing my comments

**Review by ansermino — APPROVED** [review]

**Pull request merged**